src/passes/InstrumentMemory.cpp

```diff
@@ -138,20 +161,132 @@ struct InstrumentMemory : public WalkerPass<PostWalker<InstrumentMemory>> {
       target, {builder.makeConst(int32_t(id)), curr->value}, curr->value->type);
   }
 
+  void visitStructGet(StructGet* curr) {
+    Builder builder(*getModule());
+    Name target;
+    if (curr->type == Type::i32) {
+      target = struct_get_val_i32;
+    } else if (curr->type == Type::i64) {
+      target = struct_get_val_i64;
+    } else if (curr->type == Type::f32) {
+      target = struct_get_val_f32;
+    } else if (curr->type == Type::f64) {
+      target = struct_get_val_f64;
+    } else {
+      return; // TODO: other types, unreachable, etc.
+    }
+    replaceCurrent(builder.makeCall(
+      target, {builder.makeConst(int32_t(id++)), curr}, curr->type));
```

Comment:
In `load_TYPE`/`struct_get_val_TYPE`/`array_get_val_TYPE` methods, the `value` parameter seems to be the address of the expression itself (`Load`/`StructGet`/`ArrayGet`). But In `store_TYPE`/`struct_set_TYPE`/`array_set_TYPE` methods, the `value` parameter corresponds to the value to store and there is no parameter for the expression itself (`Store`/`StructSet`/`ArraySet`).

- What is the use of logging the address of expression themselves?
- Why are they logged only for `load`/`struct_get`/`array_get`s but not for `store`/`struct_set`/`array_set`s?

Reply:
> What is the use of logging the address of expression themselves?

Do you mean the address of the Binaryen IR instruction? I don't think that is logged. We pass `curr` because we have the instruction as a child of the new one.

Overall, in both load and store, the value is the value at that address (in the case of store, after the store). We also have the address, but that is separate. A load turns into this (from the existing test):
```wat
   (call $load_val_i32
    (i32.const 1)
    (i32.load8_s
     (call $load_ptr
      (i32.const 1)
      (i32.const 1)
      (i32.const 0)
      (i32.const 0)
     )
    )
   )
```
We call `load_ptr` on the pointer, to log it. Then we do the load (using a normal wasm load). Then we call `load_val` on the loaded value, to log that.

Store has the same: a call to `store_ptr` and `store_val` for the pointer and value.

(It may be easier to read the existing test output than the code.)

Reply:
Ah I see, I was confused. Thanks.